spark/src/main/scala/com/gu/tableversions/spark/VersionedDataset.scala

```diff
@@ -117,24 +119,24 @@ object VersionedDataset {
   /**
     * Write the given partitioned dataset, storing each partition in the associated path.
     */
-  private[spark] def writeVersionedPartitions[T](dataset: Dataset[T], partitionPaths: Map[Partition, URI]): Unit = {
-    // This is a slow and inefficient implementation that writes each partition in sequence,
-    // we can look into a more performant solution later.
-
-    def filteredForPartition(partition: Partition): Dataset[T] =
-      partition.columnValues.foldLeft(dataset) {
-        case (filteredDataset, partitionColumn) =>
-          filteredDataset.where(s"${partitionColumn.column.name} = '${partitionColumn.value}'")
-      }
+  private[spark] def writeVersionedPartitions[T](
+      dataset: Dataset[T],
+      table: TableDefinition,
+      version: Version,
+      partitionPaths: Map[Partition, URI])(implicit spark: SparkSession): Unit = {
 
-    val datasetsWithPaths: Map[Dataset[T], URI] =
-      partitionPaths.map { case (partition, path) => filteredForPartition(partition) -> path }
+    spark.sparkContext.hadoopConfiguration.set("fs.versioned.version", version.label)
 
-    datasetsWithPaths.foreach {
-      case (datasetForPartition, partitionPath) =>
-        datasetForPartition.write
-          .parquet(partitionPath.toString) // TODO: Take in format parameter. Or a dataset writer?
+    VersionedFileSystem.partitionMappings(partitionPaths.keys.toList) foreach {
+      case (k, v) => spark.sparkContext.hadoopConfiguration.set(k, v)
     }
+
+    val partitions = table.partitionSchema.columns.map(_.name)
+
+    dataset.write
+      .mode(SaveMode.Append)
```

Comment:
I don't think want `Append` mode here, the default behaviour of writing but blowing up if files already exist at the given location would be better as it might catch errors? Not a biggie as it shouldn't make a difference when everything works well!